pkg/splunk/enterprise/indexercluster.go

```diff
@@ -118,23 +122,168 @@ func ApplyIndexerCluster(client splcommon.ControllerClient, cr *enterprisev1.Ind
+		if !mock {
+			return err
+		}
+	}
+
+	// Set cluster master maintenance mode
+	if enable {
+		cr.Status.ClusterMasterMaintenanceMode = true
+	} else {
+		cr.Status.ClusterMasterMaintenanceMode = false
+	}
+
+	return nil
+}
+
+// ApplyIdxcSecret checks if any of the indexer's have a different idxc_secret from namespace scoped secret and changes it
+func ApplyIdxcSecret(mgr *indexerClusterPodManager, n int32, mock bool) error {
+	// Get namespace scoped secret
+	namespaceSecret, err := splutil.ApplyNamespaceScopedSecretObject(mgr.c, mgr.cr.GetNamespace())
+	if err != nil {
+		return err
+	}
+
+	scopedLog := log.WithName("ApplyIdxcSecret").WithValues("Desired replicas", n, "IdxcSecretChanged", mgr.cr.Status.IndexerSecretChanged, "NamespaceSecretResourceVersion", mgr.cr.Status.NamespaceSecretResourceVersion, "mock", mock)
+
+	// If namespace scoped secret revision is the same ignore
+	if len(mgr.cr.Status.NamespaceSecretResourceVersion) == 0 {
+		// First time, set resource version in CR
+		mgr.cr.Status.NamespaceSecretResourceVersion = namespaceSecret.ObjectMeta.ResourceVersion
+		return nil
+	} else if mgr.cr.Status.NamespaceSecretResourceVersion == namespaceSecret.ObjectMeta.ResourceVersion {
+		// If resource version hasn't changed don't return
+		return nil
+	}
+
+	scopedLog.Info("Namespaced scoped secret revision has changed")
+
+	// Retreive idxc_secret password from secret data
+	nsIdxcSecret := string(namespaceSecret.Data["idxc_secret"])
+
+	// Loop over all indexer pods and get individual pod's idxc password
+	for i := int32(0); i <= n-1; i++ {
+		// Get Indexer's name
+		indexerPodName := GetSplunkStatefulsetPodName(SplunkIndexer, mgr.cr.GetName(), i)
+
+		// Retreive idxc_secret password from Pod
+		indIdxcSecret, err := splutil.GetSpecificSecretTokenFromPod(mgr.c, indexerPodName, mgr.cr.GetNamespace(), "idxc_secret")
+		if err != nil {
+			return fmt.Errorf("Couldn't retreive idxc_secret from secret data %s", err.Error())
+		}
+
+		// If idxc secret is different from namespace scoped secret change it
+		if indIdxcSecret != nsIdxcSecret {
+			scopedLog.Info("idxc Secret different from namespace scoped secret")
+
+			// Enable maintenance mode
+			if len(mgr.cr.Status.IndexerSecretChanged) == 0 && !mgr.cr.Status.ClusterMasterMaintenanceMode {
+				err = SetCmMaintenanceMode(mgr.c, mgr.cr, true, mock)
+				if err != nil {
+					return err
+				}
+				scopedLog.Info("Set Cm in maintenance mode")
+			}
+
+			// If idxc secret already changed, ignore
+			if i < int32(len(mgr.cr.Status.IndexerSecretChanged)) {
+				if mgr.cr.Status.IndexerSecretChanged[i] {
+					continue
+				}
+			}
+
+			// Get client for indexer Pod
+			idxcClient := mgr.getClient(i)
+
+			// Change idxc secret key
+			err = idxcClient.SetIdxcSecret(nsIdxcSecret)
+			if err != nil {
+				return err
+			}
+			scopedLog.Info("Changed idxc secret")
+
+			// Restart splunk instance on pod
+			err = idxcClient.RestartSplunk()
+			if err != nil {
+				return err
```

Comment:
Generic question: For any of these error condtions, what is the trigger for us to re-enter the next reconcile call?

Reply:
The error is carried all the way back to ApplyIndexerCluster's call to mgr.Update [here](https://github.com/splunk/splunk-operator/blob/develop/pkg/splunk/enterprise/indexercluster.go#L113)